client-java/src/main/java/uk/gov/gchq/palisade/client/Download.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2018-2021 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.palisade.client;
+
+import java.io.InputStream;
+import java.util.Optional;
+
+/**
+ * Represents a download that has been fetched. No data is transferred until
+ * {@code #getInputStream()} is called.
+ *
+ * @since 0.5.0
+ */
+public interface Download {
+
+    /**
+     * Returns an {@code InputStream} which contains the stream of bytes for the
+     * file to be downloaded
+     *
+     * @return the input stream of bytes
+     */
+    InputStream getInputStream();
+
+    /**
+     * Returns the size of the resource in bytes or -1 if not known. This can be
+     * useful for progress.
+     *
+     * @return the size of the resource in bytes or -1 if not known
+     */
+    default int getLength() {
+        return -1;
+    }
```

Comment:
I've modified the client (and added tests) to take into account the `Content-Length` header. If the server knows it (which it might if it's a file, it can add the header to the response.

Reply:
It will still likely never know.
Say, the file has 10 records, each of these records various lengths.
The data-service will never pre-process the files with rules, they will be applied at the time of a read request.
Given this, we don't know if the rules say "redact all but the first record" or "given a tree of employee-manager relations, redact any employee more than two steps away from the user's employee record" or... all sorts.
Then the only indication we can give of the length is how far through the file we are (the read pointer is, say, x% of the way through the file)
Now, if these records are all of the same length, the client can easily work out how many records were in the file before redaction, and where redacted records were, if they read through the stream one record at a time and monitor how the content-length changes.

All-in-all, big security risk that defeats the point of the whole system.

Reply:
This has been removed. It turns out that the header `Content-Length` is mutually exclusive with `Transfer-Encoding: chunked`, which kind of made sense.